tests/unit/test_connection.py

```diff
@@ -445,3 +444,51 @@ def test_task_in_response_body_without_202_status(self, mock_do_http):
         # verify the result
         self.assertEqual(mockedTaskBody, testTask)
         self.assertEqual(mockedTaskBody, testBody)
+
+    @mock.patch.object(connection, 'do_http')
+    def test_do_rest_call_with_304_status(self, mock_do_http):
+
+        mockedResponse = type('mockResponse', (), {'status': 304})()
+
+        mock_do_http.return_value = (mockedResponse, '{ "body": "test" }')
+
+        (testTask, testBody) = self.connection._connection__do_rest_call('PUT',
+                                                                         '/rest/test',
+                                                                         '{ "body": "test" }',
+                                                                         None)
+
+        self.assertIsNone(testTask)
+        self.assertEqual(testBody, {"body": "test"})
+
+    @mock.patch.object(connection, 'do_http')
+    def test_do_rest_call_with_304_status_and_invalid_json(self, mock_do_http):
+
+        mockedResponse = type('mockResponse', (), {'status': 304})()
+
+        mock_do_http.return_value = (mockedResponse, 111)
+
+        (testTask, testBody) = self.connection._connection__do_rest_call('PUT',
+                                                                         '/rest/test',
+                                                                         111,
+                                                                         None)
+
+        self.assertIsNone(testTask)
+        self.assertEqual(testBody, 111)
+
+    @mock.patch.object(connection, 'get_connection')
+    def test_download_to_stream(self, mock_get_conn):
+
+        mock_conn = mock.Mock()
+        # First attempt: Error, second attempt: successful connection
+        mock_get_conn.side_effect = [BadStatusLine(0), mock_conn]
```

Comment:
I suggest mock the sleep function to run the tests a bit faster. We are doing it on the oneview-ansible project, you just need to add something like:

```python
        self.sleep_patch = patch('time.sleep')
        self.sleep_patch.start()
        self.sleep_patch.return_value = None
```